src/librbd/Journal.h

```diff
@@ -168,6 +168,24 @@ class Journal {
     return m_work_queue;
   }
 
+protected:
+  // ensure nothing using class resources (called just before
+  // the class object is deleted). nothing special is done here
+  // expect for checking that we are in a sane state before
+  // getting detroyed.
+  void wait_for_completion() override {
```

Comment:
What is this protecting against? Can't all of this logic just be the first things in the class destructor?

Reply:
The object could still be in use after the context callback is completed unless the only thing followed after `->complete()` is a `put()`. Do you suggest going this way instead? Some parts in `librbd::Journal` assumes that the object would be destructed at a later point (after the currently executing thread finishes execution).

Locking used in this routine is just to make sure other critical sections have completed execution.

Reply:
I guess I still don't get it -- ```Journal::wait_for_completion``` looks exactly like the code that was moved from the destructor. Running ```delete this``` would run the destructor first so the result should be the same. Do you have an example of how it could still be in use but this protects it? The ```put()``` call should always be after the current thread of execute is done using the object.

Reply:
The destructor would run concurrently (in another thread) with the routine that invoked `->complete()` and the routine might still be unfinished (probably holding locks and other resources -- https://github.com/ceph/ceph/blob/master/src/librbd/Journal.cc#L612 -- this is just one instance, there are other instances which do much more that just returning after completing the context callback)

`wait_for_completion()` acquires the mutex and ensures that we're in a sane state before getting deleted. Technically, this could have been done in the destructor too.

Reply:
In that case, just acquire and release the mutex in the destructor.

Reply:
Wouldn't having completion semantics be much more cleaner. I thought acquiring mutexes in dtors is generally not a good approach.

Reply:
The clean approach would be for the ```put()``` to not be called until after the lock is released.